### ISSUE TITLE
Output of gradcam is all red!

### ISSUE THREAD

**Issue by ansariminhaj** [jacobgil/pytorch-grad-cam#108]
for data in testloader:
    images, labels = data[0].to(device), data[1].to(device)
    print(images.shape)
    img_tensor=images[1].unsqueeze_(0) 
    img_label = labels[1]
    break  

cam = GradCAM(model=net1_loaded, target_layer=net1_loaded.conv8, use_cuda=True)
target_category = 0
grayscale_cam = cam(input_tensor=img_tensor, target_category=target_category)
grayscale_cam = grayscale_cam[0, :]
print(grayscale_cam)
visualization = show_cam_on_image(img, grayscale_cam)

plt.axis('off')
plt.imshow(visualization)
plt.show()

When I print grayscale_cam all I get is a matrix of nans! The input (img_tensor) is obtained directly from the data loader class (testloader) so it is already preprocessed. The model I'm using is my own custom binary classifier CNN.

![s2](https://user-images.githubusercontent.com/46193078/124511674-b2425a00-dda4-11eb-8f7a-a0e1f01bfe4e.JPG)

**Comment by Aslyamovt** [comment]
It seems like this function 

```
cam = self.get_cam_image(input_tensor, target_category, 
            activations, grads, eigen_smooth)
```
[https://github.com/jacobgil/pytorch-grad-cam/blob/c009adac84ec83565c88c692a13bd3581e7c24eb/pytorch_grad_cam/base_cam.py#L76](url)

returns negtive values for your image then it turns into zero here

`cam = np.maximum(cam, 0)`

[https://github.com/jacobgil/pytorch-grad-cam/blob/c009adac84ec83565c88c692a13bd3581e7c24eb/pytorch_grad_cam/base_cam.py#L79](url)

and turns into NaN during normalization by max value

`img = img / np.max(img)`

[https://github.com/jacobgil/pytorch-grad-cam/blob/c009adac84ec83565c88c692a13bd3581e7c24eb/pytorch_grad_cam/base_cam.py#L85](url)

Still don't know what to do with that (I'm not a developer of this repo) but it seems like we should check this function
[https://github.com/jacobgil/pytorch-grad-cam/blob/c009adac84ec83565c88c692a13bd3581e7c24eb/pytorch_grad_cam/base_cam.py#L40](url)

**Comment by ansariminhaj** [comment]
Yeah. I np.max does go to zero. I have five convolution layers. It seems to work on the fourth one, but the fifth one (conv2d-13) gives problems.

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
================================================================
            Conv2d-1         [-1, 16, 224, 224]             448
           Dropout-2         [-1, 16, 224, 224]               0
         MaxPool2d-3         [-1, 16, 112, 112]               0
            Conv2d-4         [-1, 32, 112, 112]           4,640
           Dropout-5         [-1, 32, 112, 112]               0
         MaxPool2d-6           [-1, 32, 56, 56]               0
            Conv2d-7           [-1, 64, 56, 56]          18,496
           Dropout-8           [-1, 64, 56, 56]               0
         MaxPool2d-9           [-1, 64, 28, 28]               0
           Conv2d-10          [-1, 128, 28, 28]          73,856
          Dropout-11          [-1, 128, 28, 28]               0
        MaxPool2d-12          [-1, 128, 14, 14]               0
           Conv2d-13          [-1, 256, 14, 14]         295,168
          Dropout-14          [-1, 256, 14, 14]               0
        MaxPool2d-15            [-1, 256, 7, 7]               0
           Linear-16                 [-1, 4096]      51,384,320
          Dropout-17                 [-1, 4096]               0
           Linear-18                  [-1, 512]       2,097,664
           Linear-19                    [-1, 2]           1,026

**Comment by Aslyamovt** [comment]
Oh, you have 2-class classifier too. Grad-cam seems to work incorrect for only 2 classes. My classifier (densenet-161) almost always returns negative values on last conv layer (should return positives for grad cam to work).
For my classifier I also found out, that expected value of output on last conv weighted by gradient biased to towards negative values.
So, I think that I could solve my problem by addind some bias before

`cam = np.maximum(cam, 0)`

https://github.com/jacobgil/pytorch-grad-cam/blob/c009adac84ec83565c88c692a13bd3581e7c24eb/pytorch_grad_cam/base_cam.py#L79

**Comment by ansariminhaj** [comment]
Interesting. Do you also have a two-class network? I tried GRAD CAM with Resnet from torchvision and it seems to work for that. Did you solve the problem by adding a bias?

**Comment by Aslyamovt** [comment]
Yes, I've modified vanilla torchvision densenet-161 for satellite two-class classification. Torchvision networks pretrained on ImageNet dataset with 1000 classes (which more than 2). Seems like, if you have only 2 network could somehow classify images by sign of response on last convlutional layer and if you have more than 2 classes it have to learn generate positive responses for each class if image belongs to such class.

Bias seems to be helpful. I have calculated expected values, expected min values and expected max values of cams 
https://github.com/jacobgil/pytorch-grad-cam/blob/c009adac84ec83565c88c692a13bd3581e7c24eb/pytorch_grad_cam/base_cam.py#L76
for both zero and one predictions on my validation dataset. According to my experiments expected min values for one or zero predictions could be good bias values.

I also notised that output of `show_cam_on_image` is BGR image (OpenCV format). If you want to visualise it with plt you have to send ` use_rgb = True` to this function

**Comment by rlczddl** [comment]
I encountered a similar problem, how to add bias？
